fish-rust/src/env/mod.rs

```diff
@@ -5,4 +5,49 @@ pub mod var;
 
 pub use env_ffi::EnvStackSetResult;
 pub use environment::*;
+use std::ffi::CString;
+use std::sync::atomic::{AtomicBool, AtomicUsize};
 pub use var::*;
+
+/// Limit `read` to 100 MiB (bytes, not wide chars) by default. This can be overriden with the
+/// `fish_read_limit` variable.
+pub const DEFAULT_READ_BYTE_LIMIT: usize = 100 * 1024 * 1024;
+
+/// The actual `read` limit in effect, defaulting to [`DEFAULT_READ_BYTE_LIMIT`] but overridable
+/// with `$fish_read_limit`.
+#[no_mangle]
+pub static READ_BYTE_LIMIT: AtomicUsize = AtomicUsize::new(DEFAULT_READ_BYTE_LIMIT);
+
+/// The curses `cur_term` TERMINAL pointer has been set up.
+#[no_mangle]
+pub static CURSES_INITIALIZED: AtomicBool = AtomicBool::new(false);
+
+/// Does the terminal have the "eat new line" glitch.
+#[no_mangle]
+pub static TERM_HAS_XN: AtomicBool = AtomicBool::new(false);
+
+mod ffi {
+    extern "C" {
+        pub fn setenv_lock(
+            name: *const libc::c_char,
+            value: *const libc::c_char,
+            overwrite: libc::c_int,
+        );
+        pub fn unsetenv_lock(name: *const libc::c_char);
```

Comment:
these should be private, same for other extern C decls below

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:48 + -18
+    extern "C" {
+        /// The ncurses `cur_term` TERMINAL pointer.
+        pub static mut cur_term: *const core::ffi::c_void;
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.

we can use RelaxedAtomicBool instead of AtomicBool

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:51 + -18
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(

this no longer leaks. Presumably this is fine

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:143 + -18
+    pub fn dispatch(&self, key: &wstr, vars: &EnvStack) {
+        if let Some(named) = self.named_table.get(key) {
+            (named)(key, vars);
+        }
+        if let Some(anon) = self.anon_table.get(key) {

I don't think we can translate wcs2zstring(env_var_name) as env_var_name.to_string() because the latter does not know about our encoding scheme for codepoints that lie outside the valid Unicode ranges. 
We should use `wcs2zstring` which returns the CString we need.
Same for var.to_string().

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:168 + -18
+    } else {
+        unsetenv_lock(&name);
+    }
+
+    extern "C" {

missing the "overwriting default" FLOGF

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:309 + -18
+        USE_POSIX_SPAWN.store(false, Ordering::Relaxed);
+    } else if let Some(var) = vars.get(L!("fish_use_posix_spawn")) {
+        let use_posix_spawn =
+            var.is_empty() || crate::wcstringutil::bool_from_string(&var.as_string());
+        USE_POSIX_SPAWN.store(use_posix_spawn, Ordering::Relaxed);

The C++ code didn't update READ_BYTE_LIMIT in the error case

𑁍 5ec2fc1688916039f00716f7d3b353654abd4e12 fish-rust/src/env_dispatch.rs:539 + -41
+                    warning,
+                    wgettext!("Could not set up terminal using the fallback terminal type: "),
+                    term,
+                );
+            }

we could also add some tiny helpers in `compat.c`